File: phase2/loop/erdos1208/verify_quadratic4108373_cm_prefix_count_lock.cpp
#include <algorithm>
#include <cassert>
#include <cmath>
#include <iomanip>
#include <iostream>
#include <vector>

// Exhaustive floating-point broad sweep for the all-square/all-useful
// norm-prefix family over Q(sqrt(4108373)).  The Python companion certifies
// the exact ray rank and recomputes the closest cells with Decimal arithmetic.

namespace {

constexpr int kD = 4108373;
constexpr int kNormLimit = 2500000;
constexpr int kTMin = 50;
constexpr int kTMax = 600;
constexpr int kMaximumDepth = 8;
constexpr long double kAlpha = 0.49368647L;
constexpr long double kPackingConstant = 11978.0L / 10863.0L;

std::vector<int> primes_up_to(int limit) {
  std::vector<bool> is_prime(limit + 1, true);
  is_prime[0] = is_prime[1] = false;
  for (int p = 2; 1LL * p * p <= limit; ++p) {
    if (!is_prime[p]) continue;
    for (int q = p * p; q <= limit; q += p) is_prime[q] = false;
  }
  std::vector<int> primes;
  for (int n = 2; n <= limit; ++n) if (is_prime[n]) primes.push_back(n);
  return primes;
}

int kronecker_odd(int discriminant, int prime) {
  int a = discriminant % prime;
  if (a == 0) return 0;
  int n = prime, answer = 1;
  while (a != 0) {
    while ((a & 1) == 0) {
      a >>= 1;
      const int residue = n & 7;
      if (residue == 3 || residue == 5) answer = -answer;
    }
    std::swap(a, n);
    if ((a & 3) == 3 && (n & 3) == 3) answer = -answer;
    a %= n;
  }
  return n == 1 ? answer : 0;
}

std::vector<int> odd_prime_ideal_norms(const std::vector<int>& primes) {
  std::vector<int> degree_one, inert_squares;
  for (int p : primes) {
    if (p == 2) continue;
    const int symbol = kronecker_odd(kD, p);
    if (symbol == 0) {
      degree_one.push_back(p);
    } else if (symbol == 1) {
      degree_one.push_back(p);
      degree_one.push_back(p);
    } else if (1LL * p * p <= kNormLimit) {
      inert_squares.push_back(p * p);
    }
  }
  std::vector<int> ideals;
  std::merge(degree_one.begin(), degree_one.end(), inert_squares.begin(),
             inert_squares.end(), std::back_inserter(ideals));
  return ideals;
}

long double local_gain(int norm_q, int depth) {
  const long double x = 1.0L / (static_cast<long double>(norm_q) * norm_q);
  long double previous_sum = 1.0L, sum = 1.0L, power = 1.0L;
  for (int index = 1; index <= depth; ++index) {
    previous_sum = sum;
    power *= x;
    sum += power;
  }
  return 0.25L * (std::log1pl(1.0L / depth) + std::logl(previous_sum) -
                  std::logl(sum));
}

struct Frontier {
  std::vector<long double> cost{0.0L};
  std::vector<long double> gain{0.0L};
  std::vector<long double> slope;
  long double maximum_omitted_slope = 0.0L;

  std::pair<long double, long double> value_and_slope(long double target) const {
    assert(target > 0.0L && target < cost.back());
    const auto iterator = std::lower_bound(cost.begin(), cost.end(), target);
    const std::size_t index = iterator - cost.begin();
    assert(index > 0 && index < cost.size());
    const long double fraction =
        (target - cost[index - 1]) / (cost[index] - cost[index - 1]);
    return {gain[index - 1] + fraction * (gain[index] - gain[index - 1]),
            slope[index - 1]};
  }
};

Frontier build_frontier(const std::vector<int>& ideals, int start, int count) {
  Frontier frontier;
  std::vector<int> next(kMaximumDepth, start);
  const int end = start + count;
  for (int step = 0; step < kMaximumDepth * count; ++step) {
    int chosen = -1;
    long double chosen_slope = -1.0L;
    for (int depth_index = 0; depth_index < kMaximumDepth; ++depth_index) {
      if (next[depth_index] == end) continue;
      const int q = ideals[next[depth_index]];
      const long double cost = 0.5L * std::logl(q);
      const long double slope = local_gain(q, depth_index + 1) / cost;
      if (slope > chosen_slope) {
        chosen = depth_index;
        chosen_slope = slope;
      }
    }
    assert(chosen >= 0);
    const int q = ideals[next[chosen]++];
    const long double cost = 0.5L * std::logl(q);
    const long double gain = local_gain(q, chosen + 1);
    if (!frontier.slope.empty())
      assert(chosen_slope <= frontier.slope.back() + 1e-18L);
    frontier.cost.push_back(frontier.cost.back() + cost);
    frontier.gain.push_back(frontier.gain.back() + gain);
    frontier.slope.push_back(chosen_slope);
  }
  for (int index = start; index < end; ++index) {
    const int q = ideals[index];
    const long double cost = 0.5L * std::logl(q);
    frontier.maximum_omitted_slope = std::max(
        frontier.maximum_omitted_slope,
        local_gain(q, kMaximumDepth + 1) / cost);
  }
  return frontier;
}

struct Endpoint { long double margin, derivative, active_slope; };
struct Score {
  int t, d, useful;
  long double anchor;
  Endpoint left, right;
  long double omitted_slope;
};

Score score_configuration(const std::vector<int>& ideals,
                          const std::vector<long double>& prefix_log_cost,
                          int t) {
  const int d = t - 2;
  const int useful = (d * d - 1) / 4 - (d + 1) - t;
  assert(useful > 0 && t + useful <= static_cast<int>(ideals.size()));
  const long double log_rd =
      0.5L * std::logl(kD) + 0.25L * prefix_log_cost[t];
  const Frontier frontier = build_frontier(ideals, t, useful);

  auto endpoint = [&](long double anchor, int scale) {
    const long double w = scale * anchor;
    const auto [value, slope] = frontier.value_and_slope(2.0L * kAlpha * w);
    const long double exponent =
        2.0L * (2.0L * kAlpha - 1.0L) * w - log_rd;
    const long double ratio = std::expl(exponent) / kPackingConstant;
    const long double rhs = std::logl(kPackingConstant) + log_rd +
        (2.0L - 4.0L * kAlpha) * w + std::log1pl(ratio);
    const long double derivative =
        2.0L * kAlpha * scale * slope -
        (2.0L - 4.0L * kAlpha) * scale -
        2.0L * (2.0L * kAlpha - 1.0L) * scale * ratio / (1.0L + ratio);
    return Endpoint{value - rhs, derivative, slope};
  };

  const long double maximum_anchor =
      0.999L * frontier.cost.back() / (4.0L * kAlpha);
  long double low = maximum_anchor / 10000.0L, high = maximum_anchor;
  auto difference = [&](long double anchor) {
    return endpoint(anchor, 1).margin - endpoint(anchor, 2).margin;
  };
  bool found = false;
  long double previous = low, previous_difference = difference(previous);
  for (int step = 1; step <= 400; ++step) {
    const long double current = low + (high - low) * step / 400;
    const long double current_difference = difference(current);
    if (previous_difference * current_difference <= 0.0L) {
      low = previous;
      high = current;
      found = true;
      break;
    }
    previous = current;
    previous_difference = current_difference;
  }
  assert(found);
  long double low_difference = difference(low);
  for (int iteration = 0; iteration < 100; ++iteration) {
    const long double middle = (low + high) / 2.0L;
    const long double middle_difference = difference(middle);
    if (low_difference * middle_difference <= 0.0L) {
      high = middle;
    } else {
      low = middle;
      low_difference = middle_difference;
    }
  }
  const long double anchor = (low + high) / 2.0L;
  const Endpoint left = endpoint(anchor, 1), right = endpoint(anchor, 2);
  assert(std::fabsl(left.margin - right.margin) < 1e-10L);
  assert(left.derivative > 1e-5L && right.derivative < -1e-5L);
  assert(frontier.maximum_omitted_slope <
         std::min(left.active_slope, right.active_slope));
  return {t, d, useful, anchor, left, right,
          frontier.maximum_omitted_slope};
}

}  // namespace

int main() {
  const auto primes = primes_up_to(kNormLimit);
  const auto ideals = odd_prime_ideal_norms(primes);
  assert(ideals.size() > 90000);
  std::vector<long double> prefix_log_cost(ideals.size() + 1, 0.0L);
  for (std::size_t index = 0; index < ideals.size(); ++index)
    prefix_log_cost[index + 1] =
        prefix_log_cost[index] + std::logl(ideals[index]);

  std::vector<Score> scores;
  for (int t = kTMin; t <= kTMax; ++t)
    scores.push_back(score_configuration(ideals, prefix_log_cost, t));
  std::sort(scores.begin(), scores.end(),
            [](const Score& left, const Score& right) {
              return left.left.margin > right.left.margin;
            });
  assert(scores.size() == 551);
  assert(scores[0].t == 217 && scores[0].left.margin > 0.0L);
  assert(scores[1].t == 215 && scores[1].left.margin < -0.003L);
  assert(scores[2].t == 219 && scores[2].left.margin < -0.010L);
  assert(std::count_if(scores.begin(), scores.end(), [](const Score& score) {
           return score.left.margin >= 0.0L;
         }) == 1);

  std::cout << std::setprecision(18);
  std::cout << "range/count: 50..600 / " << scores.size() << "\n";
  for (int index = 0; index < 12; ++index) {
    const Score& score = scores[index];
    std::cout << index + 1 << " T=" << score.t << " d=" << score.d
              << " N=" << score.useful << " margin=" << score.left.margin
              << " anchor=" << score.anchor << " derivatives="
              << score.left.derivative << "," << score.right.derivative
              << "\n";
  }
  std::cout << "all-square/all-useful broad prefix count: UNIQUE T=217\n";
}

File: phase2/loop/erdos1208/verify_quadratic4108373_cm_prefix_count_lock.py
#!/usr/bin/env python3
"""Exact wrapper for the D=4108373 broad norm-prefix count lock.

The C++ companion exhausts T=50,...,600 under the favorable all-useful
relaxation.  This wrapper certifies the class/ray rank, weighted GS
bookkeeping, rational packing constants, and the closest excluded and retained
cells with high-precision Decimal arithmetic.
"""

from __future__ import annotations

from decimal import Decimal, getcontext
from fractions import Fraction
from pathlib import Path
import shutil
import subprocess
import sys
import tempfile


HERE = Path(__file__).resolve().parent
sys.path.insert(0, str(HERE))
import verify_hostile_quadratic821453_cm as base  # noqa: E402
import verify_quadratic4108373_cm_structural_lock as structural  # noqa: E402


D = 4_108_373
T_MIN = 50
T_MAX = 600
WINNER = 217
RUNNER_UP = 215
ALPHA_LOCK = Decimal("0.49368647")
C_LOWER = Fraction(11_978, 10_863)
C_UPPER = Fraction(71_603, 64_935)


def configure() -> None:
    base.D = D
    base.C = (D - 1) // 4
    base.T_COUNT = WINNER
    base.GENERATOR_RANK = WINNER - 2
    base.USEFUL_COUNT = 11_123
    base.ALPHA = ALPHA_LOCK
    base.SAFE_C = C_UPPER
    base.elementary.FIELD_DISCRIMINANT = D
    base.elementary.OMEGA_CONSTANT = (D - 1) // 4
    base.elementary.RAMIFIED_IDEAL_COUNT = WINNER
    structural.configure()


def evaluate(log_rd, frontier, packing_constant: Fraction, anchor: Decimal):
    constant = Decimal(packing_constant.numerator) / Decimal(
        packing_constant.denominator
    )
    output = []
    for endpoint in (1, 2):
        scale = Decimal(endpoint)
        w = scale * anchor
        value, index, fraction, slope = base.fast_envelope(
            frontier, 2 * ALPHA_LOCK * w
        )
        exponent = 2 * (2 * ALPHA_LOCK - 1) * w - log_rd
        ratio = exponent.exp() / constant
        rhs = (
            constant.ln()
            + log_rd
            + (2 - 4 * ALPHA_LOCK) * w
            + (1 + ratio).ln()
        )
        margin = value - rhs
        derivative = (
            2 * ALPHA_LOCK * scale * slope
            - (2 - 4 * ALPHA_LOCK) * scale
            - 2
            * (2 * ALPHA_LOCK - 1)
            * scale
            * ratio
            / (1 + ratio)
        )
        output.append((margin, derivative, slope, index, fraction))
    return output


def equal_endpoint_certificate(ideals, count: int, constant: Fraction):
    generator_rank = count - 2
    useful_count = (
        (generator_rank * generator_rank - 1) // 4
        - (generator_rank + 1)
        - count
    )
    ramified = ideals[:count]
    useful = ideals[count : count + useful_count]
    log_rd, frontier, maximum_omitted = base.prepare_endpoint_data(
        ramified, useful
    )

    low = Decimal("30000")
    high = Decimal("50000")

    def difference(anchor: Decimal) -> Decimal:
        records = evaluate(log_rd, frontier, constant, anchor)
        return records[0][0] - records[1][0]

    low_difference = difference(low)
    high_difference = difference(high)
    assert low_difference * high_difference < 0
    for _ in range(110):
        middle = (low + high) / 2
        middle_difference = difference(middle)
        if low_difference * middle_difference <= 0:
            high = middle
        else:
            low = middle
            low_difference = middle_difference
    anchor = (low + high) / 2
    records = evaluate(log_rd, frontier, constant, anchor)
    assert abs(records[0][0] - records[1][0]) < Decimal("1e-28")
    assert records[0][1] > Decimal("0.004")
    assert records[1][1] < Decimal("-0.012")
    assert maximum_omitted < min(record[2] for record in records)
    return generator_rank, useful_count, anchor, records


def exact_rank_and_gs_audit() -> tuple[int, int]:
    units, reference, raw = structural.exact_colored_ideals()
    assert reference == (2, 11, 4, 12)
    assert len(raw) >= T_MAX

    # The first nonprincipal ideal occurs at index two, so Cl_S[2] is killed
    # throughout the audited interval.  The columns consist of the two global
    # units and one S-unit column for each selected prime ideal (the reference
    # column is the generator of R_0^2).  Full rank is reached by T=6 and is
    # then monotone.  We nevertheless check every audited prefix explicitly.
    ranks = [
        base.gf2_rank(units + [row[4] for row in raw[:count]])
        for count in range(T_MIN, T_MAX + 1)
    ]
    assert set(ranks) == {4}
    assert base.gf2_rank(units + [row[4] for row in raw[:5]]) < 4
    assert base.gf2_rank(units + [row[4] for row in raw[:6]]) == 4

    for count in range(T_MIN, T_MAX + 1):
        generator_rank = count - 2
        maximum_relations = (generator_rank * generator_rank - 1) // 4
        useful_count = maximum_relations - (generator_rank + 1) - count
        assert useful_count > 0
        point = Fraction(2, generator_rank)
        assert (
            1
            - generator_rank * point
            + maximum_relations * point * point
        ) < 0
    return min(ranks), max(ranks)


def packing_constant_audit() -> None:
    sqrt_three_lower = Fraction(265, 153)
    assert sqrt_three_lower * sqrt_three_lower < 3
    fifth = Fraction(1, 5)
    atan_fifth_upper = sum(
        (-1) ** index * fifth ** (2 * index + 1) / (2 * index + 1)
        for index in range(5)
    )
    q = Fraction(1, 239)
    pi_upper = 16 * atan_fifth_upper - 4 * (q - q**3 / 3)
    assert pi_upper < Fraction(355, 113)
    assert 2 * sqrt_three_lower / Fraction(355, 113) == C_LOWER

    sqrt_three_upper = Fraction(1_351, 780)
    assert sqrt_three_upper * sqrt_three_upper > 3
    atan_fifth_lower = sum(
        (-1) ** index * fifth ** (2 * index + 1) / (2 * index + 1)
        for index in range(4)
    )
    pi_lower = 16 * atan_fifth_lower - 4 * q
    assert pi_lower > Fraction(333, 106)
    assert 2 * sqrt_three_upper / Fraction(333, 106) == C_UPPER


def run_broad_companion() -> str:
    source = Path(__file__).with_suffix(".cpp")
    compiler = shutil.which("c++")
    if compiler is None:
        raise RuntimeError("a C++17 compiler is required")
    with tempfile.TemporaryDirectory(prefix="verify_4108373_prefix_") as folder:
        executable = Path(folder) / "verify"
        subprocess.check_call(
            [
                compiler,
                "-std=c++17",
                "-O3",
                "-Wall",
                "-Wextra",
                "-pedantic",
                str(source),
                "-o",
                str(executable),
            ]
        )
        output = subprocess.check_output([str(executable)], text=True)
    assert "range/count: 50..600 / 551" in output
    assert "all-square/all-useful broad prefix count: UNIQUE T=217" in output
    assert "2 T=215 d=213 N=10913" in output
    return output


def main() -> None:
    getcontext().prec = 100
    configure()
    packing_constant_audit()
    ray_range = exact_rank_and_gs_audit()
    broad_output = run_broad_companion()

    primes = base.elementary.prime_sieve(180_000)
    ideals = base.elementary.prime_ideals(primes, 180_000)
    runner = equal_endpoint_certificate(ideals, RUNNER_UP, C_LOWER)
    winner = equal_endpoint_certificate(ideals, WINNER, C_UPPER)
    assert runner[0:2] == (213, 10_913)
    assert winner[0:2] == (215, 11_123)
    assert max(record[0] for record in runner[3]) < Decimal("-0.003")
    assert min(record[0] for record in winner[3]) > Decimal("0.0015")

    print(broad_output, end="")
    print("exact ray ranks T=50..600:", ray_range)
    print("runner T=215 / d / N / anchor:", *runner[:3])
    print("runner favorable-lower-C records:", runner[3])
    print("winner T=217 / d / N / anchor:", *winner[:3])
    print("winner adverse-upper-C records:", winner[3])
    print("D=4108373 broad prefix-count lock: CERTIFIED")


if __name__ == "__main__":
    main()
